### PR TITLE
Support omniscidb conda-forge feedstock builds with embedding feature [Linux-only]

### PR DESCRIPTION
**Pull request by pearu**
This PR includes:
- determine the omniscidb root path from the location of libDBEngine.so
- include a complete set of thrift files to installation
- use OMNISCI_ROOT_PATH environment variable to determine the root path for building Python omniscidbe extension module
- Python extension module `dbe` is renamed to `omniscidbe`

This PR is used in https://github.com/conda-forge/omniscidb-feedstock/pull/39 to build the following conda-forge packages:
- omniscdb-common, linux only
- omniscidb, cpu and cuda builds, linux only
- omniscidbe, cpu and cuda builds, linux only
- pyomniscidb, python versions 3.6, 3.7, 3.8, linux, windows, os
- pyomniscidbe, cpu and cuda builds, python version 3.6, 3.7, 3.8, linux only

### CODE REVIEW

**Comment by pearu** [comment]
@igalink @Garra1980 @vnlitvinov, please review.

**Comment by pearu** [comment]
@igalink @Garra1980 @vnlitvinov, please review.

File: Embedded/setup.in.py [hunk 1]
```diff
@@ -24,8 +34,8 @@
         "@CMAKE_SOURCE_DIR@/ThirdParty/rapidjson",
         "@CMAKE_SOURCE_DIR@/Distributed/os",
     ],
-    library_dirs=pa.get_library_dirs() + ["@CMAKE_CURRENT_BINARY_DIR@", "."],
-    runtime_library_dirs=pa.get_library_dirs() + ["$ORIGIN/../../"],
+    library_dirs=pa.get_library_dirs() + ["@CMAKE_CURRENT_BINARY_DIR@", ".", omniscidbe_root],
+    runtime_library_dirs=pa.get_library_dirs() + ["$ORIGIN/../../", omniscidbe_root],
```
Comment (vnlitvinov) [inline Embedded/setup.in.py:38]: why is this needed again?

**Review by vnlitvinov — APPROVED** [review]
LGTM, with one small question

File: Embedded/setup.in.py (hunk 1 above)
Comment (pearu) [reply Embedded/setup.in.py:38]: This is required for building Python module omniscidbe from conda recipe. Notice that `libDBEngine.so` is provided by another conda package (`omniscidbe`) and is not available as `Embedded/`libDBEngine.so` that setup.py otherwise assumes.

There are actually two possible approaches to specify the path to `libDBEngine.so`:
1. `$CONDA_PREFIX/lib`
2. `$OMNISCI_ROOT_PATH/lib`

In this PR, option 2. is used as a more general one - it would be functional when using some non-conda environment.

File: CMakeLists.txt [hunk 2]
```diff
@@ -854,6 +855,8 @@ install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/ThirdParty/licenses" DESTINATION
 
 # OmniSciTypes.h local includes (for UDF/UDTF)
 install(FILES Shared/funcannotations.h DESTINATION "Shared/" COMPONENT "include")
+install(FILES Shared/InlineNullValues.h DESTINATION "Shared/" COMPONENT "include")
```
Comment (Garra1980) [inline CMakeLists.txt:858]: Is it specific for DBEngine changes?

Comment (pearu) [reply CMakeLists.txt:858]: nope, these are the header files that OmniSciTypes.h include.

File: OSDependent/Unix/omnisci_path.cpp [hunk 3]
```diff
@@ -26,10 +26,30 @@
...
+  const struct link_map* link_map = 0;
+  const int ret = dlinfo(handle, RTLD_DI_LINKMAP, &link_map);
+  CHECK_EQ(ret, 0);
+  CHECK(link_map);
+  /* Despite the dlinfo man page claim that l_name is absolute path,
+     it is so only when the location path to the library is absolute,
+     say, as specified in LD_LIBRARY_PATH. */
+  boost::filesystem::path abs_exe_dir(boost::filesystem::absolute(
+      boost::filesystem::canonical(std::string(link_map->l_name))));
+#else
```
Comment (Garra1980) [inline OSDependent/Unix/omnisci_path.cpp:52]: So, no Windows support so far?

File: CMakeLists.txt (hunk 2 above)
Comment (Garra1980) [reply CMakeLists.txt:858]: I'm just wondering should this go as standalone PR?

File: OSDependent/Unix/omnisci_path.cpp (hunk 3 above)
Comment (pearu) [reply OSDependent/Unix/omnisci_path.cpp:52]: AFAIK, not yet. In case I am mistaken, it is better to land this PR for Linux-only because creating conda packages of omniscidb for windows is a non-trivial task on its own.

Comment (Garra1980) [reply OSDependent/Unix/omnisci_path.cpp:52]: Agree, can we reflect this somehow, eg in PR's name?

File: CMakeLists.txt [hunk 4]
```diff
@@ -338,6 +338,7 @@ if(ENABLE_DBE)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
   set(ENABLE_STANDALONE_CALCITE ON)
   add_definitions("-DENABLE_EMBEDDED_DATABASE")
+  add_definitions("-DDBEngine_LIBNAME=\"${CMAKE_SHARED_LIBRARY_PREFIX}DBEngine${CMAKE_SHARED_LIBRARY_SUFFIX}\"")
```
Comment (Garra1980) [inline CMakeLists.txt:341]: What is this for ?Do we use DBEngine_LIBNAME somewhere?

File: CMakeLists.txt (hunk 2 above)
Comment (pearu) [reply CMakeLists.txt:858]: Adding the missing header files as well as thrift files are simple bug fixes that are indirectly related to DBE packaging support (the omniscidb-feedstock would require these fixes anyway), and do not deserve a separate PR, IMHO.
If there is a particular need for the separation of these changesets then we can do it.

File: CMakeLists.txt (hunk 4 above)
Comment (pearu) [reply CMakeLists.txt:341]: Yes, this is used in
https://github.com/omnisci/omniscidb/pull/647/files#diff-572dc5a7d5354cdeee74fa14ede04679a0fad4824c1153947f8d0e10e34e09e1R41

Comment (Garra1980) [reply CMakeLists.txt:341]: Ah, right, my bad

File: OSDependent/Unix/omnisci_path.cpp (hunk 3 above)
Comment (pearu) [reply OSDependent/Unix/omnisci_path.cpp:52]: Sure, done.

Btw, currently, all omniscidb conda packages, except pyomniscidb, are Linux-only.

Comment (Garra1980) [reply OSDependent/Unix/omnisci_path.cpp:52]: Yeah, I think this is ok for now

File: CMakeLists.txt (hunk 2 above)
Comment (Garra1980) [reply CMakeLists.txt:858]: No problem, I just worried a bit how omnisci is being built now without this changes :)

**Comment by Garra1980** [comment]
I'm ok once python tests pass

**Review by Garra1980 — APPROVED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (Garra1980) [reply CMakeLists.txt:858]: Great then, thanks for clarification

**Comment by pearu** [comment]
> I'm ok once python tests pass

Locally, the three Python tests in `Embedded/test` pass ok.
These tests are now enabled also in https://github.com/conda-forge/omniscidb-feedstock/pull/39 but no confirmation yet.

File: CMakeLists.txt (hunk 4 above)
Comment (igalink) [reply CMakeLists.txt:341]: I don’t think this is necessary because different OSDependent/Unix/omnisci_path.cpp files are used for different OS.

File: Embedded/setup.in.py (hunk 1 above)
Comment (igalink) [reply Embedded/setup.in.py:38]: Could you please clarify the file location structure for the CF package build?

File: Embedded/setup.in.py [hunk 5]
```diff
@@ -7,12 +7,22 @@
...
 import numpy as np
 import pyarrow as pa
 
 root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+# conda-forge packages omniscidbe and pyomniscidbe are built
+# separately. OMNISCI_ROOT_PATH is defined by the omniscidbe activate
+# script that determines the location of libDBEngine.so and is
+# required in linking Python extension module omniscidbe.
+omnisci_root_path = os.environ.get('OMNISCI_ROOT_PATH',
```
[the remark this answers is not in the document]
Comment (pearu) [reply Embedded/setup.in.py:20]: Sure, it is defined in omniscidbe activate script:

https://github.com/conda-forge/omniscidb-feedstock/pull/39/files#diff-950dd1a30450006d4e2daa139eb257c38e5ffa2cc5f0dabbe9b12dbf485f0d00R213

File: OSDependent/Unix/omnisci_path.cpp [hunk 6]
```diff
@@ -26,10 +26,30 @@
...
+#endif
+#include <dlfcn.h>
+#include <link.h>
+#endif
 
 namespace omnisci {
 
 std::string get_root_abs_path() {
+#ifdef ENABLE_EMBEDDED_DATABASE
+  void* const handle = dlopen(DBEngine_LIBNAME, RTLD_LAZY);
```
[the remark this answers is not in the document]
Comment (pearu) [reply OSDependent/Unix/omnisci_path.cpp:41]: To my understanding, `ENABLE_EMBEDDED_DATABASE` is effective only when building `libDBEngine.so`.

The omniscidb server itself does not load `libDBEngine.so`, hence the `dlopen` is never called in this case.

Let's consider an application that links against `libDBEngine.so` (such as the Python omniscidbe module). The `get_root_abs_path` function is defined in the `libDBEngine.so` library. When the application calls `get_root_abs_path`, it must mean that the `libDBEngine.so` library has been already loaded and the `dlopen` call here should be a no-op, IIUC.
So, the issue of loading unused library is not valid because calling `get_root_abs_path` is about using the library. Let me know if I misunderstood the issue.

If there is another way to acquire the handle of a loaded library, it could be used as well.

[the remark this answers is not in the document]
Comment (igalink) [reply OSDependent/Unix/omnisci_path.cpp:41]: The get_root_abs_path function is defined in the source code which is used both for building a libDBEngine.so and a server, utilities (for example, initdb).
Actually, libDBEngine.so is built additionally to the standard omniscidb artifacts in case of ENABLE_EMBEDDED_DATABASE option.

File: Embedded/setup.in.py (hunk 1 above)
Comment (pearu) [reply Embedded/setup.in.py:38]: Sure. All paths will be shown relative to conda environment `$CONDA_PREFIX`:

1. `omniscidb-common` defines the following paths
    ```
    opt/omnisci/bin/{calcite, omnisci}-*.jar
    opt/omnisci/QueryEngine/RuntimeFunctions.bc
    ...
    ```
2. `omniscidb` defines
    ```
    opt/omnisci/bin/{omnisci_server,omnisci_initdb,omnisql,...}
    ```
3. `omniscidbe` defines
    ```
    opt/omnisci/lib/libDBEngine.so
    lib/libDBEngine.so -> ../opt/omnisci/lib/libDBEngine.so
    ```
4. `pyomniscidbe` for Python 3.8 defines
    ```
    lib/python3.8/site-packages/omniscidbe.cpython-38-x86_64-linux-gnu.so
    ```

File: Embedded/setup.in.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (igalink) [reply Embedded/setup.in.py:20]: Thus, the definition of this variable must also be added to omnisci/scripts?

[the remark this answers is not in the document]
Comment (pearu) [reply Embedded/setup.in.py:20]: No, because `omniscidb/scripts/conda` scripts use the same build directory for all components, hence `build/Embedded/libDBEngine.so` exists when running `setup.py` 

The above does not apply to omniscidb-feedstock where `omniscidbe` and `pyomniscidbe` packages are built separately.

File: OSDependent/Unix/omnisci_path.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pearu) [reply OSDependent/Unix/omnisci_path.cpp:41]: Using `dlopen` flag `RTLD_NOLOAD` might resolve the problem.

[the remark this answers is not in the document]
Comment (igalink) [reply OSDependent/Unix/omnisci_path.cpp:41]: Yes, actually this place determines whether the get_root_abs_path is located in one of the shared objects loaded by the calling application:
if (rc) {
...
} else {
      path_len = readlink("/proc/self/exe", abs_exe_path, sizeof(abs_exe_path));
}

[the remark this answers is not in the document]
Comment (igalink) [reply OSDependent/Unix/omnisci_path.cpp:41]: > Using `dlopen` flag `RTLD_NOLOAD` might resolve the problem.

Maybe you're right. If the path generated for the libDBEngine is correct for applications that do not use this library (omnisci_server, initdb).

File: Embedded/setup.in.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (igalink) [reply Embedded/setup.in.py:20]: But in setup.py, without any conditions, you take a non-existent variable, build a path relative to it and add it to library_dirs. Are you sure there are no issues?

File: Embedded/setup.in.py (hunk 1 above)
Comment (igalink) [reply Embedded/setup.in.py:38]: Thank you. Looks good

File: Embedded/setup.in.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pearu) [reply Embedded/setup.in.py]: Fixed: `OMNISCI_ROOT_PATH` is used only when defined.

**Review by igalink — APPROVED** [review]

File: OSDependent/Unix/omnisci_path.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pearu) [reply OSDependent/Unix/omnisci_path.cpp]: Now using RTLD_NOLOAD dlopen: when handler is NULL, fallback to `/proc/self/exe` path to determine the omnisci root path.

**Comment by Garra1980** [comment]
Just to check my understanding - say, I'm a developer and want to test fresh DBEngine functionality. what would be a proper way to install modified omniscidb*/pyomniscidb* components?

**Comment by pearu** [comment]
> Just to check my understanding - say, I'm a developer and want to test fresh DBEngine functionality. what would be a proper way to install modified omniscidb*/pyomniscidb* components?

To install cpu-only omniscidbe packages, use
```
conda install -c conda-forge pyomniscidbe
```
To install cuda-enabled omniscidbe packages, use
```
conda install -c conda-forge "pyomniscidbe=*=*cuda"
```

**Comment by Garra1980** [comment]
> > Just to check my understanding - say, I'm a developer and want to test fresh DBEngine functionality. what would be a proper way to install modified omniscidb*/pyomniscidb* components?
> 
> To install cpu-only omniscidbe packages, use
> 
> ```
> conda install -c conda-forge pyomniscidbe
> ```
> 
> To install cuda-enabled omniscidbe packages, use
> 
> ```
> conda install -c conda-forge "pyomniscidbe=*=*cuda"
> ```

No, I meant sort of local install, without conda-forge. If I want to test packages locally with some immediate changes.

**Comment by pearu** [comment]
> No, I meant sort of local install, without conda-forge. If I want to test packages locally with some immediate changes.

There are many approaches possible to develop omniscidb from conda environment.

To develop omniscidb, I use conda environment created from

  https://github.com/Quansight/pearu-sandbox/blob/master/conda-envs/omniscidb-dev.yaml

that is activated using the following script:

https://github.com/Quansight/pearu-sandbox/blob/master/working-envs/activate-omniscidb-internal-dev.sh

that sets up some environment variables and displays basic development instructions.

**Comment by Garra1980** [comment]
The thing is that there are already some scripts in master branch and we successfully use those scripts for development - see https://github.com/omnisci/omniscidb/blob/master/scripts/conda/build-install-all.sh. Not sure they will work after this PR though. 
And https://github.com/omnisci/omniscidb/blob/master/scripts/mapd-deps-conda-dev-env.yml helps build conda env for development (not quite correct I think)
Anyway I think we should provide (and document somewhere) the clear way of omnisci developing using conda env. Probably using scripts that are already committed. What do you think?

**Comment by pearu** [comment]
> The thing is that there are already some scripts in master branch and we successfully use those scripts for development - see https://github.com/omnisci/omniscidb/blob/master/scripts/conda/build-install-all.sh. Not sure they will work after this PR though.

I don't see why the `build-install-all.sh` would not work with this PR. If it breaks something, it should be fixed.

> And https://github.com/omnisci/omniscidb/blob/master/scripts/mapd-deps-conda-dev-env.yml helps build conda env for development (not quite correct I think)
> Anyway I think we should provide (and document somewhere) the clear way of omnisci developing using conda env. Probably using scripts that are already committed. What do you think?

I think that `mapd-deps-conda-dev-env.yml` and `build-install-all.sh` is just another equivalent way to develop omniscidb from conda environment. In the end, I think it does not matter which instruction sets to use, however, my experience is that the instructions need to be maintained and kept up-to-date at all times by developers using these actively.

I use my yaml and development scripts mainly because keeping them up-to-date is efficient and matches my work style. For instance, I use [direnv](https://direnv.net/) a lot, keep the development system minimal, and use different environments for different tasks/projects, etc, all that to facilitate working with many projects at the same time. So, many details in my scripts might be irrelevant for other developers and could be simplified.

That said, I agree that it would be nice to have a document describing the basic development workflow to ease jumping into the omniscidb development process.

**Comment by Garra1980** [comment]
> > The thing is that there are already some scripts in master branch and we successfully use those scripts for development - see https://github.com/omnisci/omniscidb/blob/master/scripts/conda/build-install-all.sh. Not sure they will work after this PR though.
> 
> I don't see why the `build-install-all.sh` would not work with this PR. If it breaks something, it should be fixed.

Right, right. I just meant we haven't tried it yet. Will check and fix in case of any issues

> 
> > And https://github.com/omnisci/omniscidb/blob/master/scripts/mapd-deps-conda-dev-env.yml helps build conda env for development (not quite correct I think)
> > Anyway I think we should provide (and document somewhere) the clear way of omnisci developing using conda env. Probably using scripts that are already committed. What do you think?
> 
> I think that `mapd-deps-conda-dev-env.yml` and `build-install-all.sh` is just another equivalent way to develop omniscidb from conda environment. In the end, I think it does not matter which instruction sets to use, however, my experience is that the instructions need to be maintained and kept up-to-date at all times by developers using these actively.
> 
> I use my yaml and development scripts mainly because keeping them up-to-date is efficient and matches my work style. For instance, I use [direnv](https://direnv.net/) a lot, keep the development system minimal, and use different environments for different tasks/projects, etc, all that to facilitate working with many projects at the same time. So, many details in my scripts might be irrelevant for other developers and could be simplified.
> 
> That said, I agree that it would be nice to have a document describing the basic development workflow to ease jumping into the omniscidb development process.

We will start handle this, I think
